src/libmongoc/src/mongoc/mongoc-ts-pool-private.h

```diff
@@ -1,97 +1,310 @@
 
 /**
- * @brief Allocate a buffer of data suitable for the given pool.
+ * @brief Destroy an item that was created by a pool
+ *
+ * Instead of returning to the pool, the item will be destroyed and deallocated.
+ *
+ * @param item A pointer returned by `mongoc_ts_pool_get` or
+ * `mongo_ts_pool_get_existing`.
+ */
+void
+mongoc_ts_pool_drop (void *item);
+
+/**
+ * @brief Visit each element of the pool, optionally pruning items.
+ *
+ * @note While this visit function is executing, all pool operations are
+ * blocked on all threads. Attempting to return/get/drop items in the pool from
+ * within the visit callback will deadlock.
+ *
+ * The visit function will be called as:
+ *
+ *    visit(pool_item, pool_userdata, visit_userdata)
+ *
+ * If the visitor function returns non-zero for an item, then that item will be
+ * removed from the pool.
+ *
+ * @param pool The pool of items to visit
+ * @param visit_userdata Optional additional context for the visitor
+ * @param visitor The visit function
  *
- * @returns A pointer to uninitialized memory of the element size of the pool.
- * The memory MUST be initialized before the returned pointer is given to
- * `mongo_ts_pool_free_item` or `mongo_ts_pool_push`.
  */
-extern void *
-mongoc_ts_pool_alloc_item (mongoc_ts_pool pool);
+void
+mongoc_ts_pool_visit_each (mongoc_ts_pool *pool,
+                           void *visit_userdata,
+                           int (*visitor) (void *item,
+                                           void *pool_userdata,
+                                           void *visit_userdata));
 
 /**
- * @brief Destroy an item that was created with `mongo_ts_pool_alloc_item`.
+ * @brief Declare a thread-safe pool type that contains elements of a specific
+ * type. Wraps a `mongoc_ts_pool`.
  *
- * @param item A pointer returned by `mongo_ts_pool_alloc_item` or
- * `mongo_ts_pool_try_pop`
+ * @param ElementType The type of object contained in the pool
+ * @param PoolName The name of the pool type. All methods of the pool will be
+ * prefixed by this name.
+ * @param Constructor A function that constructs new elements for the pool, or
+ * `NULL`
+ * @param Destructor An element destructor function, or `NULL`
+ * @param PrunePredicate A function that checks whether elements should be
+ * dropped from the pool, or `NULL`
  */
-extern void
-mongoc_ts_pool_free_item (mongoc_ts_pool pool, void *item);
+#define MONGOC_DECL_SPECIAL_TS_POOL(ElementType,                              \
+                                    PoolName,                                 \
+                                    UserDataType,                             \
+                                    Constructor,                              \
+                                    Destructor,                               \
+                                    PrunePredicate)                           \
+   typedef struct PoolName {                                                  \
```

Comment:
Removing the tag name on the struct makes it invalid to use as an explicit tagged name. This isn't a huge problem, but is very confusing when for `typedef struct {} foo;`, then `foo` and `struct foo` no longer name the same type. Alternatively, it may be useful to remove the `typedef` and allow the macro-user to declare the `typedef` themselves.

Reply:
Ah I mistakenly assumed this was unintentional since `struct PoolName` was not referred to. Resolving.